### PR TITLE
Add MarketplaceHooks and MarketFee traits

### PR DESCRIPTION
**Pull request by HashWarlock**
## Description
Currently the Market pallet has no way to check if the type of NFT can be sold (i.e. Stakepool v2 NFTs vs PhalaWorld Shell NFTs). To resolve this, there are an introduction to 2 new traits called `MarketplaceHooks` and `MarketFee` which will hold the standard market fee percentage constant in the `MarketFee` trait and the `MarketplaceHooks` will implement 3 functions called `calculate_market_fee(amount)` `calculate_royalty_fee(amount, royalty_fee)` and `can_list_or_buy_in_market(collections_id, nft_id)` that will help prevent NFTs from being listed that are not eligible for the marketplace.

## Targets
- [x] add new traits `MarketFee` and `MarketplaceHooks`
- [x] implement `Default` for the traits in runtime
- [x] access control function in `MarketplaceHooks` trait
- [x] run integration tests
- [x] calculate new weights if needed

### CODE REVIEW

File: pallets/rmrk-core/src/functions.rs [hunk 1]
```diff
@@ -569,6 +569,9 @@ impl<T: Config>
 		nft_id: T::ItemId,
 		budget: &dyn Budget,
 	) -> DispatchResultWithPostInfo {
+		// Check Lock to prevent locked NFTs from being burned. Owner must unlock the NFT before
+		// burning.
+		ensure!(!Pallet::<T>::is_locked(collection_id, nft_id), pallet_uniques::Error::<T>::Locked);
```
Comment (Leechael) [inline pallets/rmrk-core/src/functions.rs:574]: Nice design here! Can we add a collection-level lock?

Comment (HashWarlock) [reply pallets/rmrk-core/src/functions.rs:574]: We can do this with the `pallet_uniques` function.
```
		/// Disallow further unprivileged transfers for a whole collection.
		///
		/// Origin must be Signed and the sender should be the Freezer of the `collection`.
		///
		/// - `collection`: The collection to be frozen.
		///
		/// Emits `CollectionFrozen`.
		///
		/// Weight: `O(1)`
		#[pallet::call_index(9)]
		#[pallet::weight(T::WeightInfo::freeze_collection())]
		pub fn freeze_collection(
			origin: OriginFor<T>,
			collection: T::CollectionId,
		) -> DispatchResult {
			let origin = ensure_signed(origin)?;

			Collection::<T, I>::try_mutate(collection, |maybe_details| {
				let details = maybe_details.as_mut().ok_or(Error::<T, I>::UnknownCollection)?;
				ensure!(origin == details.freezer, Error::<T, I>::NoPermission);

				details.is_frozen = true;

				Self::deposit_event(Event::<T, I>::CollectionFrozen { collection });
				Ok(())
			})
		}
```

**Comment by HashWarlock** [comment]
@jasl let me know if good to merge. Then can update the cargo in https://github.com/Phala-Network/khala-parachain/pull/254

**Review by jasl — APPROVED** [review]

**Comment by jasl** [comment]
MERGED!

**Pull request merged**